module/jobqueue/consumer.go

```diff
@@ -0,0 +1,309 @@
+package jobqueue
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/storage"
+)
+
+type Worker interface {
+	Run(job storage.Job)
+}
+
+type Consumer struct {
+	sync.Mutex
+	log zerolog.Logger
+
+	// Storage
+	jobs     storage.Jobs             // storage to read jobs from
+	progress storage.ConsumerProgress // storing the last processed job, so that we can resume after restarting
+	worker   Worker                   // defines how jobs will be processed
+
+	// Config
+	maxProcessing int64 // max number of jobs to be processed concurrently
+
+	// State Variables
+	running bool // a signal to control whether to start processing more jobs. Useful for waiting
+	// until the workers are ready
+	isChecking *atomic.Bool // allow only one process checking job processable
+	// are ready, and stop when shutting down.
+
+	processedIndex   int64
+	processings      map[int64]*jobStatus // keep track of the status of each on going job
+	processingsIndex map[JobID]int64      // lookup the index of the job, useful when fast forwarding the
```

Comment:
As discussed channel based could work. Just that the lock-based implementation is easier to understand, especially we need to support async job (worker waiting for the engine to report a job is done), and also be able to lookup the next unprocessed index (we anyway need to lock the consumer in order to update the `processedIndex`